### DISCUSSION TITLE
How to use the Github cache action in a containerized workflow?

### DISCUSSION THREAD

**Discussion by obfischer** [community/community#23607] [New to GitHub]
<p>I would like to run a Github Actions workflow in a container and reuse an existing cache.</p>
<pre><code class="lang-auto">name: Build Application

on:
  workflow_dispatch:

jobs:
  build-native-application:
    name: Build native application
    runs-on: ubuntu-20.04
    container:
      image: ghcr.io/graalvm/graalvm-ce:21.3.0
    steps:
      - name: Checkout
        uses: actions/checkout@v2
      - name: Cache local downloads
        uses: actions/cache@v2
        with:
          path: ~/downloads
          key: ${{ runner.os }}-downloads-{{ hashFiles('image/*.sha512') }}
      - name: Cache local Maven repository
        uses: actions/cache@v2
        with:
          path: ${{ github.workspace }}/.m2/repository
          key: ${{ runner.os }}-maven-${{ hashFiles('**/pom.xml') }}
</code></pre>
<p>The worfklow runs fine and in the log I can see the also that the requested caches have been found, although I am not able to find them anywhere in my running container. I asume, that the caches will be created in the runner and not in the container.</p>
<p>Does someone know my blind point or knows a way to reuse a cache inside a Docker containere in a workflow?</p>

**Comment by AllanOricil** [comment]
<p>I may be with the same problem. I can see git restoring my m2 cache, but it has size equal to 0. So I think the m2 folder is in the runner, not in the container. Can someone help us? Thank in advance.</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/3X/c/c/cceacd76c1ad8dd806d3f6bce78cff15aed98455.jpeg" data-download-href="/uploads/short-url/teMsU6lVxb3KCfvNqMC9CqKkRHn.jpeg?dl=1" title="image" rel="noopener nofollow ugc"><img src="https://user-images.githubusercontent.com/55927613/181078181-ef93ad3d-49f3-452e-9d7c-bfb16e6324e5.jpeg" alt="image" data-base62-sha1="teMsU6lVxb3KCfvNqMC9CqKkRHn" width="690" height="210" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/c/c/cceacd76c1ad8dd806d3f6bce78cff15aed98455_2_690x210.jpeg, https://aws1.discourse-cdn.com/github/optimized/3X/c/c/cceacd76c1ad8dd806d3f6bce78cff15aed98455_2_1035x315.jpeg 1.5x, https://aws1.discourse-cdn.com/github/optimized/3X/c/c/cceacd76c1ad8dd806d3f6bce78cff15aed98455_2_1380x420.jpeg 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/c/c/cceacd76c1ad8dd806d3f6bce78cff15aed98455_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use href="#far-image"></use></svg><span class="filename">image</span><span class="informations">1512×461 89.2 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use href="#discourse-expand"></use></svg>
</div></a></div>
<pre><code class="lang-auto">name: Clean Test Workflow Template

on:
  workflow_call:
    inputs:
      MULESOFT_APPS:
        required: true
        type: string
    secrets:
      MULE_NEXUS_USERNAME:
        required: true
      MULE_NEXUS_PASSWORD:
        required: true
      MULE_USERNAME:
        required: true
      MULE_PASSWORD:
        required: true

jobs:
  clean-test:
    name: clean test
    runs-on: ubuntu-latest
    container:
      image: ghcr.io/vodafoneis/mulesoft-build-image:main
      credentials:
        username: ${{ github.actor }}
        password: ${{ secrets.GITHUB_TOKEN }}

    strategy:
      matrix:
        MULESOFT_APP: ${{ fromJson(inputs.MULESOFT_APPS) }}

    env:
      MULE_NEXUS_USERNAME: ${{ secrets.MULE_NEXUS_USERNAME }}
      MULE_NEXUS_PASSWORD: ${{ secrets.MULE_NEXUS_PASSWORD }}
      MULE_USERNAME: ${{ secrets.MULE_USERNAME }}
      MULE_PASSWORD: ${{ secrets.MULE_PASSWORD }}

    defaults:
      run:
        working-directory: ${{ matrix.MULESOFT_APP }}

    steps:
      - uses: actions/checkout@v2

      - name: cache mvn modules
        uses: actions/cache@v2
        env:
          cache-name: cache-mvn-dependencies
        with:
          path: ~/.m2
          key: ${{ matrix.MULESOFT_APP }}-${{ hashFiles('pom.xml') }}
          restore-keys: |
            ${{ matrix.MULESOFT_APP }}-
            ${{ matrix.MULESOFT_APP }}

      - name: print app name
        run: echo ${{ matrix.MULESOFT_APP }}

      - name: clean test
        run: /scripts/clean_build_and_test.sh -s ../../settings.xml

</code></pre>

**Comment by AllanOricil** [comment, marked as answer]
<p><a class="mention" href="/obfischer">@obfischer</a></p>
<p>I just found out that mvn dependencies are at /root/.m2/repository and not at ~/.m2/repository (this resolves to github/home/.m2/repository wich is empty)<br>
To find this, I run this command: mvn help:evaluate -Dexpression=settings.localRepository</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/3X/c/4/c40d8aa890a65ca52654157e21d24446dfb53689.png" data-download-href="/uploads/short-url/rYmB0Y5JGhh9jaCOdfHb5fAObfH.png?dl=1" title="image" rel="noopener nofollow ugc"><img src="https://user-images.githubusercontent.com/55927613/181078182-52a69a57-593b-4cc4-a410-b030e7c7cedf.png" alt="image" data-base62-sha1="rYmB0Y5JGhh9jaCOdfHb5fAObfH" width="395" height="500" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/c/4/c40d8aa890a65ca52654157e21d24446dfb53689_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use href="#far-image"></use></svg><span class="filename">image</span><span class="informations">439×555 65.1 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use href="#discourse-expand"></use></svg>
</div></a></div>

**Comment by airtower-luna** [comment]
<aside class="quote no-group" data-username="AllanOricil" data-post="2" data-topic="219514">
<div class="title">
<div class="quote-controls"></div>
 AllanOricil:</div>
<blockquote>
<p>So I think the m2 folder is in the runner, not in the container.</p>
</blockquote>
</aside>
<p>I haven’t used Maven in Actions, but I have run into issues caused by paths being different inside a job container. You can see what gets mounted where in the Docker command line in the “Starting job container” group of the “Initialize containers” step, and then adjust for it.</p>

**Comment by AllanOricil** [comment]
<p>Here is my workflow restoring mvn cache when I used /root/.m2/repository</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/3X/6/4/64de5e8f67abb8be08b5f3fc7893b6b1f5c1657b.png" data-download-href="/uploads/short-url/eokaCe6TKx4WAkgtoHIKkQoO73R.png?dl=1" title="image" rel="noopener nofollow ugc"><img src="https://user-images.githubusercontent.com/55927613/181078185-e2e05e5d-1a13-41cb-9fbe-3d279851ffb9.png" alt="image" data-base62-sha1="eokaCe6TKx4WAkgtoHIKkQoO73R" width="690" height="161" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/6/4/64de5e8f67abb8be08b5f3fc7893b6b1f5c1657b_2_690x161.png, https://aws1.discourse-cdn.com/github/optimized/3X/6/4/64de5e8f67abb8be08b5f3fc7893b6b1f5c1657b_2_1035x241.png 1.5x, https://aws1.discourse-cdn.com/github/original/3X/6/4/64de5e8f67abb8be08b5f3fc7893b6b1f5c1657b.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/6/4/64de5e8f67abb8be08b5f3fc7893b6b1f5c1657b_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use href="#far-image"></use></svg><span class="filename">image</span><span class="informations">1127×263 82.5 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use href="#discourse-expand"></use></svg>
</div></a></div>
<p>And here you can see that mvn did not download dependencies again</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/3X/5/5/55238ac0952b803c1f0123d66f1a320be411b6b5.jpeg" data-download-href="/uploads/short-url/c9aJt5CTrgFxetA2pwwmN9jTM5D.jpeg?dl=1" title="image" rel="noopener nofollow ugc"><img src="https://user-images.githubusercontent.com/55927613/181078186-52169686-96cc-494e-b01f-1d676796c33d.jpeg" alt="image" data-base62-sha1="c9aJt5CTrgFxetA2pwwmN9jTM5D" width="535" height="500" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/5/5/55238ac0952b803c1f0123d66f1a320be411b6b5_2_535x500.jpeg, https://aws1.discourse-cdn.com/github/optimized/3X/5/5/55238ac0952b803c1f0123d66f1a320be411b6b5_2_802x750.jpeg 1.5x, https://aws1.discourse-cdn.com/github/original/3X/5/5/55238ac0952b803c1f0123d66f1a320be411b6b5.jpeg 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/5/5/55238ac0952b803c1f0123d66f1a320be411b6b5_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use href="#far-image"></use></svg><span class="filename">image</span><span class="informations">1000×933 163 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use href="#discourse-expand"></use></svg>
</div></a></div>

**Comment by airtower-luna** [comment]
<p>Ah, that makes sense, because the commands inside the Docker container run as root. Thank you for sharing! :slightly_smiling_face:</p>

**Comment by AllanOricil** [comment]
I can't believe I forgot I had done this before hahaha
I just google it again 

But here is a more complete answer that would make ur cache step work on both container and not container based builds.

![image](https://user-images.githubusercontent.com/55927613/219540272-5eadecd5-ce26-4ad4-a665-381bc7c6f8d6.png)

**Reply by AntonioLyubchev** [reply]
Do you know how we can use the cache in a docker build? For example with 
- uses: docker/build-push-action@v5 
this action?

**Reply by AllanOricil** [reply]
> Do you know how we can use the cache in a docker build? For example with 
> - uses: docker/build-push-action@v5 
> this action?

Maybe you can restore the m2 folder and then move it to the "context" where the image is going to be built. Context is a prop in the action. I believe that if maven first look up for m2 in the same directory as the pom.xml, then goes to the user dir, then the root dir.
